test/helpers/quictestlib.c

```diff
@@ -331,7 +337,31 @@ int qtest_create_quic_objects(OSSL_LIB_CTX *libctx, SSL_CTX *clientctx,
 
 void qtest_add_time(uint64_t millis)
 {
+    if (!CRYPTO_THREAD_write_lock(fake_now_lock))
+        return;
     fake_now = ossl_time_add(fake_now, ossl_ms2time(millis));
+    CRYPTO_THREAD_unlock(fake_now_lock);
+}
+
+static OSSL_TIME qtest_get_time(void)
+{
+    OSSL_TIME ret;
+
+    if (!CRYPTO_THREAD_read_lock(fake_now_lock))
+        return ossl_time_zero();
+    ret = fake_now;
+    CRYPTO_THREAD_unlock(fake_now_lock);
+    return ret;
+}
+
+static void qtest_reset_time(void)
+{
+    if (!CRYPTO_THREAD_write_lock(fake_now_lock))
+        return;
+    fake_now = ossl_time_zero();
+    CRYPTO_THREAD_unlock(fake_now_lock);
+    /* zero time can have a special meaning, bump it */
+    qtest_add_time(1);
```

Comment:
Is this racy? Could you just ossl_time_add() inside the write lock?

Alternatively you can see this function as something that should be never used in racy contexts and drop the lock calls completely here. Using this in a racy context would mean that we need to have multiple fake_now storages per use actually because otherwise one thread could reset the fake time for another innocent thread and the time would go backwards.

Reply:
I don't think this is racy.  qtest_add_time retakes the global lock, so there aren't any read/write races strictly speaking.  It is theoretically possible that another thread might read the zero time, but given that this is during setup, there should be no other threads running